ush/python/pygw/src/pygw/schema.py

```diff
@@ -0,0 +1,891 @@
+                    "predictable values, like literal strings or ints. "
+                    '"%r" is too complex.' % (self._schema,)
+                )
+            self.default = default
+            self.key = str(self._schema)
+
+    def __hash__(self):
+        return hash(self._schema)
+
+    def __eq__(self, other):
+        return (
+            self.__class__ is other.__class__
+            and getattr(self, "default", self._MARKER) == getattr(other, "default", self._MARKER)
+            and self._schema == other._schema
+        )
+
+    def reset(self):
+        if hasattr(self._schema, "reset"):
+            self._schema.reset()
+
+
+class Hook(Schema):
+    def __init__(self, *args, **kwargs):
+        self.handler = kwargs.pop("handler", lambda *args: None)
+        super(Hook, self).__init__(*args, **kwargs)
+        self.key = self._schema
+
+
+class Forbidden(Hook):
+    def __init__(self, *args, **kwargs):
+        kwargs["handler"] = self._default_function
+        super(Forbidden, self).__init__(*args, **kwargs)
+
+    @staticmethod
+    def _default_function(nkey, data, error):
+        raise SchemaForbiddenKeyError(
+            "Forbidden key encountered: %r in %r" % (nkey, data), error)
+
+
+class Literal(object):
+    def __init__(self, value, description=None):
+        self._schema = value
+        self._description = description
+
+    def __str__(self):
+        return self._schema
+
+    def __repr__(self):
+        return 'Literal("' + self.schema + '", description="' + (self.description or "") + '")'
+
+    @property
+    def description(self):
+        return self._description
+
+    @property
+    def schema(self):
+        return self._schema
+
+
+class Const(Schema):
+    def validate(self, data, **kwargs):
+        super(Const, self).validate(data, **kwargs)
+        return data
+
+
+def _callable_str(callable_):
+    if hasattr(callable_, "__name__"):
+        return callable_.__name__
+    return str(callable_)
+
+
+def _plural_s(sized):
+    return "s" if len(sized) > 1 else ""
+
+# ----
+
+# The following functions are specific to global-workflow. The Schema
+# class module was obtained from:
+
+# https://github.com/keleshev/schema/blob/master/schema.py
+
+
+# TODO: Is there a better way to do this?
+logger = Logger("pygw.schema", colored_log=True)
+
+
+@logit(logger)
+def build_schema(yaml_path: str) -> Dict:
+    """
+    Description
+    -----------
+
+    This function parses a YAML-formatted file and defines the
+    respective schema.
+
+    Parameters
+    ----------
+
+    yaml_path: str
+
+        A Python string specifying the path to the YAML-formatted file
+        containing the schema attributes.
+
+    schema_dict: Dict
+
+        A Python dictionary containing the schema.
+
+    """
+
+    # Read the YAML-formatted file containing the schema.
+    data = yaml_file.parse_yaml(path=yaml_path)
+
+    schema_dict = {}
+    for datum in data:
+        data_dict = AttrDict(data[datum])
```

Comment:
What purpose does `AttrDict` really serve here other than ability to get attribute.
```suggestion
        data_dict = data[datum]
```